google/cloud/internal/oauth2_external_account_credentials.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/oauth2_external_account_credentials.h"
+#include "google/cloud/internal/external_account_parsing.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/internal/rest_client.h"
+#include <nlohmann/json.hpp>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+ExternalAccountCredentials::ExternalAccountCredentials(
+    ExternalAccountInfo info, HttpClientFactory client_factory, Options options)
+    : info_(std::move(info)),
+      client_factory_(std::move(client_factory)),
+      options_(std::move(options)) {}
+
+StatusOr<internal::AccessToken> ExternalAccountCredentials::GetToken(
+    std::chrono::system_clock::time_point tp) {
+  auto subject_token = (info_.token_source)(Options{});
```

Comment:
I have been punting on this because I do not know where the options should come from.  I added an item in #5915, it may turn into a separate bug because I think most credentials do this poorly (or inconsistently).